hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/cluster/strategy/ClusteringExecutionStrategy.java

```diff
@@ -67,4 +85,69 @@ protected HoodieEngineContext getEngineContext() {
+        .withMaxMemorySizeInBytes(maxMemory)
+        .withReverseReader(config.getCompactionReverseLogReadEnabled())
+        .withBufferSize(config.getMaxDFSStreamBufferSize())
+        .withSpillableMapBasePath(config.getSpillableMapBasePath())
+        .withPartition(operation.getPartitionPath())
+        .withOptimizedLogBlocksScan(config.enableOptimizedLogBlocksScan())
+        .withDiskMapType(config.getCommonConfig().getSpillableDiskMapType())
+        .withBitCaskDiskMapCompressionEnabled(config.getCommonConfig().isBitCaskDiskMapCompressionEnabled())
+        .withRecordMerger(config.getRecordMerger())
+        .withTableMetaClient(table.getMetaClient())
+        .build();
+
+    Option<HoodieFileReader> baseFileReader = StringUtils.isNullOrEmpty(operation.getDataFilePath())
+        ? Option.empty()
+        : Option.of(getBaseOrBootstrapFileReader(storageConf, bootstrapBasePath, partitionFields, operation));
+    Option<BaseKeyGenerator> keyGeneratorOp = getKeyGenerator();
+    try {
+      return new HoodieFileSliceReader(baseFileReader, scanner, readerSchemaWithMetaFields, tableConfig.getPreCombineField(), config.getRecordMerger(),
+          tableConfig.getProps(),
+          tableConfig.populateMetaFields() ? Option.empty() : Option.of(Pair.of(tableConfig.getRecordKeyFieldProp(),
+              tableConfig.getPartitionFieldProp())), keyGeneratorOp);
+    } catch (IOException e) {
+      throw new HoodieClusteringException("Error reading file slices", e);
+    }
+  }
+
+  protected ClosableIterator<HoodieRecord<T>> getRecordIteratorWithBaseFileOnly(ClusteringOperation operation) {
+    StorageConfiguration<?> storageConf = getHoodieTable().getStorageConf();
+    HoodieTableConfig tableConfig = getHoodieTable().getMetaClient().getTableConfig();
+    String bootstrapBasePath = tableConfig.getBootstrapBasePath().orElse(null);
+    Option<String[]> partitionFields = tableConfig.getPartitionFields();
+    HoodieFileReader baseFileReader = getBaseOrBootstrapFileReader(storageConf, bootstrapBasePath, partitionFields, operation);
+
+    Option<BaseKeyGenerator> keyGeneratorOp = getKeyGenerator();
+    // NOTE: Record have to be cloned here to make sure if it holds low-level engine-specific
+    //       payload pointing into a shared, mutable (underlying) buffer we get a clean copy of
+    //       it since these records will be shuffled later.
+    ClosableIterator<HoodieRecord> baseRecordsIterator;
+    try {
+      baseRecordsIterator = baseFileReader.getRecordIterator(readerSchemaWithMetaFields);
+    } catch (IOException e) {
+      throw new HoodieClusteringException("Error reading base file", e);
+    }
+    return new CloseableMappingIterator(
+        baseRecordsIterator,
+        rec -> ((HoodieRecord) rec).copy().wrapIntoHoodieRecordPayloadWithKeyGen(readerSchemaWithMetaFields, writeConfig.getProps(), keyGeneratorOp));
+  }
+
+  protected abstract Option<BaseKeyGenerator> getKeyGenerator();
```

Comment:
Let's remove the interface for `getKeyGenerator` and `getBaseOrBootstrapFileReader`, they are just utilities.

Reply:
> Let's remove the interface for `getKeyGenerator` and `getBaseOrBootstrapFileReader`, they are just utilities.

done~